src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1865,14 +1865,21 @@ private AuthenticationResult getTokenWithRefreshToken(final IWindowComponent act
         // useCache will be false when calling from acquireTokenUsingRefreshToken, and if false, need to return
         // the error through callback. If true, just return the result back to localflow. 
         if (useCache) {
-            if (result == null || StringExtensions.IsNullOrBlank(result.getAccessToken())) {
-                String errLogInfo = result == null ? "" : result.getErrorLogInfo();
-                Logger.w(TAG, "Refresh token did not return accesstoken.", request.getLogInfo()
-                        + errLogInfo, ADALError.AUTH_FAILED_NO_TOKEN);
-
-                // remove item from cache to avoid same usage of
-                // refresh token in next acquireToken call
-                removeItemFromCache(refreshItem);
+            if (result == null) {
+                Logger.w(TAG, "Received empty response from refresh token request.", request.getLogInfo(), 
+                        ADALError.AUTH_FAILED_NO_TOKEN);
+                return result;
+            } else if (StringExtensions.IsNullOrBlank(result.getAccessToken())) {
+                final String errorLogInfo = result.getErrorLogInfo();
+                Logger.e(TAG, "Refresh token request failed to return accesstoken.", 
+                        request.getLogInfo() + errorLogInfo, ADALError.AUTH_FAILED_NO_TOKEN);
+                
+                // check error code, only remove token from cache if receive invalid_grant from server
+                if (AuthenticationConstants.OAuth2ErrorCode.INVALID_GRANT.equalsIgnoreCase(result.getErrorCode())) {
+                    Logger.v(TAG, "Removing token cache for invalid_grant error returned from server.");
+                    removeItemFromCache(refreshItem);
+                }
+                
                 return result;
```

Comment:
@agamemnum We are going to expose the specific oauth error via the cause property on the existing exception.  The cause will contain a AuthenticationServerProtocolException whose ADALError property will be 'OAuth'.  There will be a protocolSpecificError string typed property that will contain the actual OAuth error code.  You can look for that and if you find the value 'invalid_grant' you will know that the account is broken and the user will need to sign-in again rather than simply retrying.